pkg/istio/servicerole/servicerole.go

```diff
@@ -53,14 +75,151 @@ func (srMgr *ServiceRoleMgr) GetServiceRoleMap() (map[string]*ServiceRoleInfo, e
+func NewConfig(namespace string, name string, spec *v1alpha1.ServiceRole) model.Config {
+
+	meta := model.ConfigMeta{
+		Type:      model.ServiceRole.Type,
+		Group:     model.ServiceRole.Group + model.IstioAPIGroupDomain,
+		Version:   model.ServiceRole.Version,
+		Namespace: namespace,
+		Name:      name,
+	}
+	return model.Config{
+		ConfigMeta: meta,
+		Spec:       spec,
+	}
+}
+
+// parseAssertionEffect parses the effect of an assertion into a supported Istio RBAC action
+func parseAssertionEffect(assertion *zms.Assertion) (string, error) {
+	if assertion == nil {
+		return "", fmt.Errorf("assertion is nil")
+	}
+	effect := assertion.Effect
+	if effect == nil {
+		return "", fmt.Errorf("assertion effect is nil")
+	}
+	if strings.ToUpper(effect.String()) != zms.ALLOW.String() {
+		return "", fmt.Errorf("effect: %s is not a supported assertion effect", effect)
+	}
+	return zms.ALLOW.String(), nil
+}
+
+// parseAssertionAction parses the action of an assertion into a supported Istio RBAC HTTP method
+func parseAssertionAction(assertion *zms.Assertion) (string, error) {
+	if assertion == nil {
+		return "", fmt.Errorf("assertion is nil")
+	}
+	method := strings.ToUpper(assertion.Action)
+	if !supportedMethods[method] {
+		return "", fmt.Errorf("method: %s is not a supported HTTP method", assertion.Action)
+	}
+	return method, nil
+}
+
+// parseAssertionResource parses the resource of an action into the service name (AccessRule constraint) and the
+// HTTP paths if specified (suffix :<path>)
+func parseAssertionResource(domainName zms.DomainName, assertion *zms.Assertion) (string, string, error) {
+
+	if assertion == nil {
+		return "", "", fmt.Errorf("assertion is nil")
+	}
+	var svc string
+	var path string
+	resource := assertion.Resource
+	parts := resourceRegex.FindStringSubmatch(resource)
+	names := resourceRegex.SubexpNames()
+	results := map[string]string{}
+	for i, part := range parts {
+		results[names[i]] = part
+	}
+
+	for name, match := range results {
+		switch name {
+		case "domain":
+			if match != string(domainName) {
+				return "", "", fmt.Errorf("resource: %s does not belong to the Athenz domain: %s", resource, domainName)
+			}
+		case "svc":
+			svc = match
+		case "path":
+			path = match
+		}
+	}
+
+	if svc == "" {
+		return "", "", fmt.Errorf("resource: %s does not specify the service using svc.<service-name> format", resource)
+	}
+	return svc, path, nil
+}
+
+// GetServiceRoleSpec returns the ServiceRoleSpec for a given Athenz role and the associated assertions
+func GetServiceRoleSpec(domainName zms.DomainName, roleName string, assertions []*zms.Assertion) (*v1alpha1.ServiceRole, error) {
+
+	rules := make([]*v1alpha1.AccessRule, 0)
+	for _, assertion := range assertions {
+		assertionRole, err := common.ParseRoleFQDN(domainName, string(assertion.Role))
+		if err != nil {
+			log.Println(err.Error())
+			continue
+		}
+
+		if assertionRole != roleName {
```

Comment:
Shouldn't the assertions already be grouped by the role at this point? Why are we doing this check?

Reply:
yep, this is extra safety check, since the actual grouping was done by athenz.model, and this one is rbac provider implementation specific, wanted to have the check explicit